ygot/pathstrings.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ygot
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/openconfig/ygot/util"
+
+	gnmipb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+// PathToString takes a gNMI Path and provides its string representation. For example,
+// the path Path{Element: []string{"one", "two", "three"} is converted to the string
+// "/one/two/three" and returned. Both the pre-0.4.0 "element"-based paths, and the
+//  >0.4.0 paths based on "elem" are supported. In the case that post-0.4.0 paths are
+// specified, keys that are specified in the path are concatenated onto the name of
+// the path element using the format [name=value]. If the path specifies both pre-
+// and post-0.4.0 paths, the pre-0.4.0 version is returned.
+func PathToString(path *gnmipb.Path) string {
+	var buf bytes.Buffer
+	buf.WriteRune('/')
+	if path.Element != nil {
+		for i, e := range path.Element {
+			buf.WriteString(e)
+			if i != len(path.Element)-1 {
+				buf.WriteRune('/')
+			}
+		}
+		return buf.String()
+	}
+
+	for i, e := range path.Elem {
+		buf.WriteString(e.Name)
+		if len(e.Key) != 0 {
+			var keys []string
+			for k := range e.Key {
+				keys = append(keys, k)
+			}
+			sort.Strings(keys)
+
+			for _, k := range keys {
+				buf.WriteString(fmt.Sprintf("[%s=%s]", k, e.Key[k]))
+			}
+		}
+		if i != len(path.Elem)-1 {
+			buf.WriteRune('/')
+		}
+	}
+	return buf.String()
+}
+
+// PathType is used to indicate a gNMI path type.
+type PathType int64
+
+const (
+	// StructuredPath represents a Path using the structured 'PathElem' message in
+	// the 'elem' field of the gNMI Path message.
+	StructuredPath PathType = iota
+	// StringSlicePath represents a Path using the 'Element' repeated string field
+	// of the gNMI path message.
+	StringSlicePath
+)
+
+// StringToPath takes an input string representing a path in gNMI, and converts
+// it to a gNMI Path message, populated with the specified path encodings.
+func StringToPath(path string, pathTypes ...PathType) (*gnmipb.Path, util.Errors) {
```

Comment:
We specifically want a func that populates both formats at the same time.
That's how we'll handle transition from Element to Elem without breaking any APIs.

This func can be changed to take only one PathType, but then add another value PathBoth (can't come up with better name) that would populate both.

Reply:
Does it have to take a PathType? i.e. could we have three functions:
StringToGNMIStringPath
StringToGNMIStructuredPath
StringToGNMIPath 
It's better to have three smaller functions than one big, complex one.

Reply:
I'm not sure I get this -- there are already two functions that do this separately. For example, we have `StringToStringSlicePath` which is essentially `StringToGNMIStringPath` -- the same thing exists for the structured path. Only in the StringToPath public API are both populated, and this is just a wrapper around the two paths.

The reason I added the `...PathType` here, was to avoid `PathBoth`, which seemed klunky compared to just a slice of the path types one wants. They're all public APIs of this package right now.